doc/development/testing.rst

```diff
@@ -0,0 +1,206 @@
+`gtest <https://github.com/google/googletest>`_ framework, whereas Python projects use
+`unittest <http://pythontesting.net/framework/unittest/unittest-introduction/>`_.
+For node level tests, `rostest <https://wiki.ros.org/rostest>`_ is used.
+Rostest uses launch files to launch the nodes under test and the actual tests.
+
+The next sections will describe the process of writing tests. For this
+tutorial we will use the :codedir:`ros_test_tutorial package <development/ros_test_tutorial>`.
+This package already includes some c++ and python code to write tests for.
+
+Writing c++ unit tests
+^^^^^^^^^^^^^^^^^^^^^^
+Tests are always located in the ``test/`` directory of a package. To create a
+test we create a new cpp file in the ``test`` directory called ``AddTest.cpp``
+with the following contents.
+
+.. code::
+
+    #include <gtest/gtest.h>
+
+    #include "ros_test_tutorial/Add.h"
+
+    TEST(AddTest, addTwoAndOne)
+    {
+        Add add;
+        ASSERT_EQ(add.add(2, 1), 3);
+    }
+
+    int main(int argc, char* argv[])
+    {
+        testing::InitGoogleTest(&argc, argv);
+        return RUN_ALL_TESTS();
+    }
+
+Feel free to experiment with writing tests. The actual test is written using
+the ``TEST`` macro.  The first argument is the test suite, the same as the
+filename in this case.  The second argument is name of the actual test. So this
+would be another example of a test.
+
+.. code::
+
+    TEST(AddTest, addNegative)
+    {
+        Add add;
+        ASSERT_EQ(add.add(0, -1), -1);
+    }
+
+In order to build and run the unit tests we must add ``rosunit`` as test
+dependency to our package. So normally we would add the following to the
+``package.xml``. However, this has already been done in highlevel
+``package.xml``.
+
+.. code::
+
+  <test_depend>rosunit</test_depend>
+
+Next we must tell cmake which tests it has to build. So add the following to the ``CMakeLists.txt``.
+
+.. code::
+
+    if(CATKIN_ENABLE_TESTING)
+        catkin_add_gtest(add_test test/AddTest.cpp)
+        target_link_libraries(add_test ros_test_tutorial ${catkin_LIBRARIES})
+    endif()
+
+Here we tell ``cmake`` to build the test and use our library. Now when
+``colcon test`` is run from the workspace root you should see tests passing.
+
+Writing c++ node tests
+^^^^^^^^^^^^^^^^^^^^^^
+
+
+Writing python unit tests
+^^^^^^^^^^^^^^^^^^^^^^^^^
+Tests for python are also always located in the ``test/`` directory. We will
+first start by writing our test. So create the ``MultiplyTest.py`` file to your
+``test/`` directory and fill it with the following contents.
+
+.. code::
+
+    #!/usr/bin/env python
+
+    import unittest
+    import rosunit
+
+    from ros_test_tutorial import multiply
+
+    class MultiplyTest(unittest.TestCase):
+
+        def test_multiply_one_and_one():
+            self.assertEqual(multiply(1, 1), 1)
+
+    if __name__ == '__main__':
+        rosunit.unitrun(PKG, 'test_multiply', MultiplyTest)
```

Comment:
What does 'test_multiply' mean? Is that just a string that is displayed? Also, how does rosunit.unitrun now which functions in MultiplyTest are testfunctions it should run? (Does it run all functions, or all that start with "test" or something)